File: Python3/581. Shortest Unsorted Continuous Subarray.py
class Solution:
    def findUnsortedSubarray(self, nums: List[int]) -> int:
        if len(nums) < 2:
            return 0
        
        stack = deque()
        left, right = len(nums)-1, 0
        
        for i, n in enumerate(nums):
            while stack and nums[stack[-1]] > n:
                left = min(left, stack.pop())
            stack.append(i)
            
        stack.clear()
        for i in range(len(nums)-1, -1, -1):
            while stack and nums[stack[-1]] < nums[i]:
                right = max(right, stack.pop())
            stack.append(i)
            
        return right-left+1 if right > left else 0

File: Java/581. Shortest Unsorted Continuous Subarray.java
class Solution {
    public int findUnsortedSubarray(int[] nums) {
        Stack<Integer> stack = new Stack<>();
        int left = nums.length-1, right = 0;
        
        for (int i=0; i<nums.length; i++) {
            while (!stack.isEmpty() && nums[stack.peek()] > nums[i]) {
                left = Math.min(left, stack.pop());
            }
            stack.push(i);
        }
        
        stack.clear();
        for (int i=nums.length-1; i>=0; i--) {
            while (!stack.isEmpty() && nums[stack.peek()] < nums[i]) {
                right = Math.max(right, stack.pop());
            }
            stack.push(i);
        }
        
        return right > left ? right-left+1 : 0;
    }
}